evm/client/rpc_client.go

```diff
@@ -1330,47 +1267,24 @@ func (r *RPCClient) wrapHTTP(err error) error {
 }
 
 // makeLiveQueryCtxAndSafeGetClients wraps makeQueryCtx
-func (r *RPCClient) makeLiveQueryCtxAndSafeGetClients(parentCtx context.Context, timeout time.Duration) (ctx context.Context, cancel context.CancelFunc, ws *rawclient, http *rawclient) {
-	ctx, cancel, _, ws, http = r.acquireQueryCtx(parentCtx, timeout)
-	return
+func (r *RPCClient) makeLiveQueryCtxAndSafeGetClients(parentCtx context.Context, timeout time.Duration) (ctx context.Context, cancel context.CancelFunc,
+	ws *rawclient, http *rawclient) {
+	ctx, cancel, _, wsClient, httpClient := r.acquireQueryCtx(parentCtx, timeout)
+	return ctx, cancel, wsClient.Load(), httpClient.Load()
 }
 
 func (r *RPCClient) acquireQueryCtx(parentCtx context.Context, timeout time.Duration) (ctx context.Context, cancel context.CancelFunc,
-	chStopInFlight chan struct{}, ws *rawclient, http *rawclient) {
-	// Need to wrap in mutex because state transition can cancel and replace the
-	// context
-	r.stateMu.RLock()
-	chStopInFlight = r.chStopInFlight
-	if r.ws != nil {
-		cp := *r.ws
-		ws = &cp
-	}
-	if r.http != nil {
-		cp := *r.http
-		http = &cp
-	}
-	r.stateMu.RUnlock()
-	ctx, cancel = makeQueryCtx(parentCtx, chStopInFlight, timeout)
-	return
-}
-
-// makeQueryCtx returns a context that cancels if:
-// 1. Passed in ctx cancels
-// 2. Passed in channel is closed
-// 3. Default timeout is reached (queryTimeout)
-func makeQueryCtx(ctx context.Context, ch services.StopChan, timeout time.Duration) (context.Context, context.CancelFunc) {
-	var chCancel, timeoutCancel context.CancelFunc
-	ctx, chCancel = ch.Ctx(ctx)
-	ctx, timeoutCancel = context.WithTimeout(ctx, timeout)
-	cancel := func() {
-		chCancel()
-		timeoutCancel()
+	chStopInFlight chan struct{}, ws atomic.Pointer[rawclient], http atomic.Pointer[rawclient]) {
```

Comment:
no need to return atomic.Pointers. You can return rawClients or remove acquireQueryCtx